### PR TITLE
feat: Add ratelimiting to OpWitness API

### LINKED ISSUE #12871: Add ratelimiting to OpWitness API
### Describe the feature

currently this rpc endpoint is unbounded

https://github.com/paradigmxyz/reth/blob/1b4048e47d42c0053ea9ba216281ceb513d70bc4/crates/optimism/rpc/src/witness.rs#L54-L54

we should add internal ratelimiting because this is fairly expensive:

https://github.com/paradigmxyz/reth/blob/1b4048e47d42c0053ea9ba216281ceb513d70bc4/crates/optimism/rpc/src/witness.rs#L78-L78

## TODO
* make function async (requires adding async_trait macro for DebugExecutionWitnessApiServer) and make this async: https://github.com/paradigmxyz/reth/blob/1b4048e47d42c0053ea9ba216281ceb513d70bc4/crates/rpc/rpc-api/src/debug.rs#L404-L404
* Introduce a new tokio semaphor
* add `Box<dyn TaskSpawner>` to https://github.com/paradigmxyz/reth/blob/1b4048e47d42c0053ea9ba216281ceb513d70bc4/crates/optimism/rpc/src/witness.rs#L78-L78 so we can spawn as blocking
needs update here https://github.com/paradigmxyz/reth/blob/1b4048e47d42c0053ea9ba216281ceb513d70bc4/crates/optimism/node/src/node.rs#L236-L236

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by stevencartavia**
should close #12871

### CODE REVIEW

File: crates/rpc/rpc-eth-api/src/helpers/pending_block.rs
Comment (mattsse) [inline crates/rpc/rpc-eth-api/src/helpers/pending_block.rs:1]: these are unrelated, please undo

File: crates/optimism/rpc/src/witness.rs [hunk 1]
```diff
@@ -22,9 +24,13 @@ pub struct OpDebugWitnessApi<Provider, EvmConfig> {
...
     /// Creates a new instance of the `OpDebugWitnessApi`.
-    pub fn new(provider: Provider, evm_config: EvmConfig) -> Self {
+    pub fn new(
+        provider: Provider,
+        evm_config: EvmConfig,
+        task_spawner: Box<dyn TaskSpawner>,
+    ) -> Self {
         let builder = OpPayloadBuilder::new(evm_config);
-        let inner = OpDebugWitnessApiInner { provider, builder };
+        let inner = OpDebugWitnessApiInner { provider, builder, task_spawner };
```
Comment (mattsse) [inline crates/optimism/rpc/src/witness.rs:33]: good start, we also want to add a new tokio Semaphor for ratelimiting

**Review by mattsse — CHANGES_REQUESTED** [review]

File: crates/rpc/rpc-eth-api/src/helpers/pending_block.rs
Comment (stevencartavia) [reply crates/rpc/rpc-eth-api/src/helpers/pending_block.rs:1]: just the `pending_block.rs` or also `block.rs` and `transactions.rs`?

File: crates/optimism/rpc/src/witness.rs (hunk 1 above)
Comment (stevencartavia) [reply crates/optimism/rpc/src/witness.rs:33]: what would be the number of permits? 

and I'm not sure if this is the correct place and way to use it :(

```
impl<Provider, EvmConfig> OpDebugWitnessApi<Provider, EvmConfig> {
    /// Creates a new instance of the `OpDebugWitnessApi`.
    pub fn new(
        provider: Provider,
        evm_config: EvmConfig,
        task_spawner: Box<dyn TaskSpawner>,
    ) -> Self {
        let builder = OpPayloadBuilder::new(evm_config);
        let semaphore = Arc::new(Semaphore::new(3));
        let inner = OpDebugWitnessApiInner { provider, builder, task_spawner, semaphore };
        Self { inner: Arc::new(inner) }
    }
}
```

Comment (mattsse) [reply crates/optimism/rpc/src/witness.rs:33]: 3 seems fine

**Review by mattsse — APPROVED** [review]
lgtm,

only did the last few remaining touchups
